webrtc/RTCPeerConnection-addIceCandidate.html

```diff
@@ -0,0 +1,531 @@
+    }));
+  }, 'Add ICE candidate after setting remote description should succeed');
+
+  promise_test(t => {
+    const pc = new RTCPeerConnection();
+
+    return pc.setRemoteDescription(sessionDesc)
+    .then(() => pc.addIceCandidate(new RTCIceCandidate({
+      candidate: candidateStr1,
+      sdpMid, sdpMLineIndex, ufrag
+    })));
+  }, 'Add ICE candidate with RTCIceCandidate should succeed');
+
+  promise_test(t => {
+    const pc = new RTCPeerConnection();
+
+    return pc.setRemoteDescription(sessionDesc)
+    .then(() => pc.addIceCandidate({
+      candidate: candidateStr1,
+      sdpMid, sdpMLineIndex, ufrag
+    }))
+    .then(() => {
+      assert_candidate_line_between(pc.remoteDescription.sdp,
+        mediaLine1, candidateLine1, mediaLine2);
+    });
+  }, 'Added candidate should be found in first media stream');
+
+  promise_test(t => {
+    const pc = new RTCPeerConnection();
+
+    return pc.setRemoteDescription(sessionDesc)
+    .then(() => pc.addIceCandidate({
+      candidate: candidateStr2,
+      sdpMid: sdpMid2,
+      sdpMLineIndex: sdpMLineIndex2,
+      ufrag: ufrag2
+    }))
+    .then(() => {
+      assert_candidate_line_after(pc.remoteDescription.sdp,
+        mediaLine2, candidateLine2);
+    });
+  }, 'Added candidate should be found in second media stream');
+
+  promise_test(t => {
+    const pc = new RTCPeerConnection();
+
+    return pc.setRemoteDescription(sessionDesc)
+    .then(() => pc.addIceCandidate({
+      candidate: candidateStr1,
+      sdpMid, sdpMLineIndex, ufrag
+    }))
+    .then(() => pc.addIceCandidate({
+      candidate: candidateStr2,
+      sdpMid: sdpMid2,
+      sdpMLineIndex: sdpMLineIndex2,
+      ufrag: ufrag2
+    }))
+    .then(() => {
+      assert_candidate_line_between(pc.remoteDescription.sdp,
+        mediaLine1, candidateLine1, mediaLine2);
+
+      assert_candidate_line_after(pc.remoteDescription.sdp,
+        mediaLine2, candidateLine2);
+    });
+  }, 'Multiple candidates should be added to their corresponding media stream');
+
+  promise_test(t => {
+    const pc = new RTCPeerConnection();
+
+    return pc.setRemoteDescription(sessionDesc)
+    .then(() => pc.addIceCandidate({
+      candidate: candidateStr1,
+      sdpMid, sdpMLineIndex, ufrag
+    }))
+    .then(() => pc.addIceCandidate({
+      candidate: '',
+      sdpMid, sdpMLineIndex,
+      ufrag
+    }))
+    .then(() => {
+      assert_candidate_line_between(pc.remoteDescription.sdp,
+        mediaLine1, candidateLine1, mediaLine2);
+
+      assert_candidate_line_between(pc.remoteDescription.sdp,
+        mediaLine1, endOfCandidateLine, mediaLine2);
+    });
+  }, 'Add with empty candidate string (end of candidate) should succeed');
+
+  // WebIDL for addIceCandidate does not allow null argument.
+  // This throws a TypeError because it is checked at WebIDL level.
+  // null can be accidentally passed from onicecandidate event handler
+  // when null is used to indicate end of candidate
```

Comment:
Actually that is wrong. I later remembered WebIDL explicitly states that if the return type of the method is a promise, it converts any error thrown during checking to promise rejection. I updated the code but forgot to remove the comment.

https://heycam.github.io/webidl/#es-operations